ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoiceNotifier.java

```diff
@@ -132,32 +128,71 @@ private void internalTerminalBlockReached(Bytes32 executionBlockHash) {
+      // current Future Payload ID builds on the correct block
+
+      if (lastFuturePayloadId.isEmpty()) {
+        // consistency check
+        throw new IllegalStateException(
+            "we cannot have lastSentForkChoiceState without lastFuturePayloadId");
       }
-      throw new IllegalStateException(
-          String.format("PayloadId not available for Beacon Block Root %s", parentBeaconBlockRoot));
+
+      return lastFuturePayloadId
+          .get()
+          .thenApply(
+              payloadId -> {
+
+                // we accept empty payload only in pre-merge
+                if (payloadId.isPresent() || parentExecutionHash.isZero()) return payloadId;
+
+                throw new IllegalStateException(
+                    String.format(
+                        "PayloadId not available for Beacon Block Root %s", parentBeaconBlockRoot));
+              });
     }
+
+    // we have no SentForkChoiceState or doesn't build on top of the right block
     if (parentExecutionHash.isZero()) {
-      // Haven't reached the merge block yet, parent has default payload
-      if (executionPayloadTerminalBlockHash.isPresent()
-          && lastSentForkChoiceHead.equals(executionPayloadTerminalBlockHash.get())) {
-        // Creating the merge block - payload ID is building on the terminal PoW block
-        return lastPayloadId;
+      // pre-merge
+
+      if (executionPayloadTerminalBlockHash.isEmpty()) {
+        // we are in pre-merge and terminal block is not reached, we can build a block with empty
+        // payload
+        return SafeFuture.completedFuture(Optional.empty());
+      }
+
+      if (allowPayloadIdOnTheFlyRetrieval) {
+        // we are in pre-merge and terminal block is reached
+        // try to obtain a payloadId now
+        final Bytes32 terminalBlockHash = executionPayloadTerminalBlockHash.get();
+        final ForkChoiceState terminalForkChoiceState =
+            new ForkChoiceState(terminalBlockHash, terminalBlockHash, Bytes32.ZERO);
+        internalForkChoiceUpdated(terminalForkChoiceState);
+        if (!lastSentForkChoiceState.equals(forkChoiceState) || lastFuturePayloadId.isEmpty()) {
+          throw new IllegalStateException();
+        }
+        return internalGetPayloadId(parentBeaconBlockRoot, false);
       }
-      // Can still use a default payload
-      return Optional.empty();
-    } else {
-      // Merge is complete so we must have a real payload, but we don't have one that matches
-      // TODO: try to obtain a payloadId on the fly?
-      throw new IllegalStateException(
-          "Payload ID required but not available for parent block root " + parentBeaconBlockRoot);
     }
+
+    // Merge is complete so we must have a real payload, but we don't have one that matches
+    // TODO: try to obtain a payloadId on the fly?
+
+    throw new IllegalStateException(
+        String.format("PayloadId not available for Beacon Block Root %s", parentBeaconBlockRoot));
```

Comment:
Probably should call `internalForkChoiceUpdated` here in the same way as above but using `parentExecutionHash` instead of `terminalBlockHash`.  Can probably extract a `requestPayloadId(Bytes32 parentExecutionBlockHash)` kind of method to share the code between the two cases.

We'd also need to use the finalized hash from the last sent fork choice update (or `ZERO` if there hasn't been one) when sending the update but that should work for both cases too.

Could do this in a follow up PR as it shouldn't happen in normal operation.